crypto/s2n_hash.c

```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
     return 0;
 }
 
-int s2n_hash_reset(struct s2n_hash_state *state)
+static int s2n_low_level_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
 {
-    return s2n_hash_init(state, state->alg);
+    memcpy_check(to, from, sizeof(struct s2n_hash_state));
+    return 0;
 }
 
-int s2n_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
+static int s2n_low_level_hash_reset(struct s2n_hash_state *state)
 {
-    memcpy_check(to, from, sizeof(struct s2n_hash_state));
+    return s2n_low_level_hash_init(state, state->alg);
+}
+
+static int s2n_low_level_hash_free(struct s2n_hash_state *state)
+{
+    return 0;
+}
+
+static int s2n_evp_hash_new(struct s2n_hash_state *state)
+{
+    notnull_check(state->digest.evp.primary.ctx = S2N_EVP_MD_CTX_NEW());
+    notnull_check(state->digest.evp.md5_secondary.ctx = S2N_EVP_MD_CTX_NEW());
+
+    return 0;
+}
+
+static int s2n_evp_hash_init(struct s2n_hash_state *state, s2n_hash_algorithm alg)
+{
+    int r;
+    switch (alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_md5(), NULL);
+        break;
+    case S2N_HASH_SHA1:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha1(), NULL);
+        break;
+    case S2N_HASH_SHA224:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha224(), NULL);
+        break;
+    case S2N_HASH_SHA256:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha256(), NULL);
+        break;
+    case S2N_HASH_SHA384:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha384(), NULL);
+        break;
+    case S2N_HASH_SHA512:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha512(), NULL);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_sha1(), NULL);
+        r &= EVP_DigestInit_ex(state->digest.evp.md5_secondary.ctx, EVP_md5(), NULL);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_INIT_FAILED);
+    }
+
+    state->alg = alg;
+
+    return 0;
+}
+
+static int s2n_evp_hash_update(struct s2n_hash_state *state, const void *data, uint32_t size)
+{
+    int r;
+    switch (state->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
```

Comment:
I thought I remembered you saying that you were going to use a wrapper-style interface (similar to how s2n_cipher is) instead of the switch-case-style. Was that still something you were considering? I personally feel like the wrapper-style is cleaner and more consistent with other places in the code base, but I could be wrong.